File: 백준/Silver/25956. 목차 세기/목차 세기.java
import java.io.*;
import java.util.*;

public class Main {
    // 스택에 저장할 항목 클래스
    static class StackItem {
        int level;  // 제목 수준
        int index;  // 해당 제목의 인덱스

        StackItem(int level, int index) {
            this.level = level;
            this.index = index;
        }
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());
        
        int[] v = new int[n+1];       // 하위 제목 개수 저장 배열
        int[] answers = new int[n]; // 결과 배열
        Deque<StackItem> stack = new ArrayDeque<>(); // ArrayDeque를 스택으로 사용 (성능상 이점)
        int before = 0;
        
        for (int i = 0; i < n; i++) {
            int tmp = Integer.parseInt(br.readLine());  // 입력
            
            if (stack.isEmpty() || stack.peek().level <= tmp) {  // 스택이 비었거나, 스택 최상단 값보다 입력 값이 크거나 같을 때
                if (before + 1 < tmp) {  // 입력값 검증
                    System.out.println("-1");
                    return;
                }
                before = tmp;  // 이전 값 갱신
                
                // 스택에 추가
                stack.push(new StackItem(tmp, i));
                
                // tmp는 tmp-1의 하위 제목
                if (tmp > 0) {
                    v[tmp - 1]++;
                }
            } else {  // 스택이 비어있지 않고, 스택 최상단 값보다 입력 값이 작을 때
                before = tmp;  // 이전 값 갱신
                
                // 스택 최상단 값이 tmp보다 크거나 같을 동안 스택에서 요소 제거
                while (!stack.isEmpty() && tmp <= stack.peek().level) {
                    StackItem item = stack.pop();
                    answers[item.index] = v[item.level]; // 결과 배열에 저장
                    v[item.level] = 0;  // 하위제목 개수 초기화
                }
                
                // 스택에 추가
                stack.push(new StackItem(tmp, i));
                
                // tmp는 tmp-1의 하위 제목
                if (tmp > 0) {
                    v[tmp - 1]++;
                }
            }
        }
        
        // 남은 스택 요소 처리
        while (!stack.isEmpty()) {
            StackItem item = stack.pop();
            answers[item.index] = v[item.level];
            v[item.level] = 0;
        }
        
        // 결과 출력
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < n; i++) {
            sb.append(answers[i]).append("\n");
        }
        System.out.print(sb);
    }
}

File: 백준/Silver/25956. 목차 세기/목차 세기.c
#include <stdio.h>

int main(void)
{
	int n;
	scanf("%d", &n);
	
	int v[n]; //ex) v[1]은 제목 1 아래의 제목 2의 개수를 의미합니다. 즉 하위제목의 개수를 저장하는 배열입니다.
	int ans[n]; //답 (대입을 이용하니 초기화 하지 않아도 됩니다)
	int stack[n][2]; //스택 (대입을 이용하니 초기화 하지 않아도 됩니다)
	int cnt = 0; //스택의 position indicator
	int before = 0;
	
	for(int i=0; i<n; i++) { //n개 입력
        v[i] = 0; //이건 초기화 해주어야 합니다.
		int tmp;
		scanf("%d", &tmp); //입력
		
		if(!cnt || stack[cnt-1][0] <= tmp) { //스택이 비어있거나, 스택에서 가장 최근 것 보다 값이 같거나 더 클 때
			if(before+1 < tmp) { //입력이 잘못되었는지를 검사
				printf("-1");
				return 0;
			} 
			before = tmp; //갱신
            
            //스택에 집어넣기
			stack[cnt][0] = tmp;
			stack[cnt++][1] = i;
            
            //tmp은 tmp-1의 하위 제목입니다.
			v[tmp-1]++;
            
		} else { //스택에 요소가 있고, 스택에서 가장 최근 것보다 값이 작을 때
			before = tmp; //갱신
            
			while(cnt>0 && tmp <= stack[cnt-1][0]) { //스택보다 tmp의 값이 클 때까지 계속 스택의 요소들을 꺼냅니다.
				--cnt;
                int tmp2 = stack[cnt][0]; //스택에서 꺼낸 값의 값(제목)
				ans[stack[cnt][1]] = v[tmp2]; //답을 출력하는 배열에 스택에서 꺼낸 값의 값(제목)의 바로 아래 하위제목의 개수를 대입
				v[tmp2] = 0; //그런 다음 바로 아래 하위제목의 개수를 0으로 바꿈
			}
            
            //스택에 집어넣기
			stack[cnt][0] = tmp;
			stack[cnt++][1] = i;
            
            //tmp은 tmp-1의 하위 제목입니다
			v[tmp-1]++;
		}
	}
	
	while(cnt--) { //스택에 남은 값 꺼내기
        int tmp2 = stack[cnt][0]; //스택에서 꺼낸 값의 값(제목)
		ans[stack[cnt][1]] = v[tmp2]; //답을 출력하는 배열에 스택에서 꺼낸 값의 값(제목)의 바로 아래 하위제목의 개수를 대입
		v[tmp2] = 0; //그런 다음 바로 아래 하위제목의 개수를 0으로 바꿈
	}
	for(int i=0; i<n; i++) printf("%d\n", ans[i]); //입력받은대로 출력
}